workflow/controller/controller.go

```diff
@@ -542,6 +535,75 @@ func (wfc *WorkflowController) runPodCleanup(ctx context.Context) {
 	}
 }
 
+func (wfc *WorkflowController) podHasStatusFinalizer(pod *apiv1.Pod) bool {
+	for _, f := range pod.ObjectMeta.Finalizers {
+		if f == common.FinalizerPodStatus {
+			return true
+		}
+	}
+	return false
+}
+
+func (wfc *WorkflowController) getPodCleanupPatch(pod *apiv1.Pod, labelPodCompleted bool) ([]byte, error) {
+	needPatch := false
+	metadata := map[string]interface{}{}
+	if labelPodCompleted {
+		if pod.Labels != nil || pod.Labels[common.LabelKeyCompleted] != "true" {
+			metadata["labels"] = map[string]string{common.LabelKeyCompleted: "true"}
+			needPatch = true
+		}
+	}
+	finalizerEnabled := os.Getenv("ARGO_POD_STATUS_CAPTURE_FINALIZER") == "true"
+	if finalizerEnabled {
+		if wfc.podHasStatusFinalizer(pod) {
+			finalizers := make([]string, 0)
+			for _, f := range pod.ObjectMeta.Finalizers {
+				if f != common.FinalizerPodStatus {
+					finalizers = append(finalizers, f)
+				}
+			}
+			metadata["finalizers"] = finalizers
+			metadata["resourceVersion"] = pod.ObjectMeta.ResourceVersion
```

Comment:
> I have found merge patch can also follow the resourceVersion check if we pass it to request. [...]

Huh, if you can do that with a patch, then I guess this works. But we need to handle possible conflicts then; the previous `Update` code used `RetryOnConflict` for that.
The `Patch` request on line 593 does not handle conflicts

Reply:
I was doing some investigation on this and popped up on some things:
1. https://github.com/kubernetes/kubernetes/issues/105146 -- deleting finalizers is the exact use-case for us.
   - More specifically that's a bug to watch out for if we were to use `$deleteFromPrimitiveList` (not sure which version of k8s it's fixed in).
   - More generically, we can use directives in a strategic merge patch to do this more cleanly. Stumbled upon this [SO answer](https://stackoverflow.com/a/64565206/3431180) answer which took me to these [SIG API Machinery docs](https://github.com/kubernetes/community/blob/94b696bef96aefba2ca9bf97029694565a84495e/contributors/devel/sig-api-machinery/strategic-merge-patch.md) that describe the directives you can use in k8s.
       - I think we should be able to use `$delete` or `$deleteFromPrimitiveList` for this.
1. We could maybe do an `Update` on the finalizer subresource? That would be less network since it's just finalizers

Reply:
1. I can find that fix in release-1.23 and later branches. As https://github.com/kubernetes/kubernetes/issues/105146#issuecomment-925330928 mentioned, merge patch + resourverVersion can do a consistent patch. Wether to use $deleteFromPrimitiveList or merger patch + resourceVersion depends on the lowest version of kubernetes that argo wants to support. 
2. As we can see from https://github.com/kubernetes/kubernetes/blob/release-1.24/pkg/registry/core/rest/storage_core.go#L280, pod have no finalizer subresource.